nova_ecs/utils.ts

```diff
@@ -47,30 +50,192 @@ export function topologicalSortList(list: Sortable[]): Sortable[] {
+        │ a1
+       ┌└>root
+       │
+       └──a2
+    */
+    const listing:T[] = [];
+    const indices = new Map<T,number>();
+    const seen = new Set<T>();
+    for (let k of graph.keys()){
+        if (seen.has(k)){
+            continue;
+        }
+        do {
+            seen.add(k);
+            indices.set(k,listing.length);
+            listing.push(k);
+            k = setDifference(graph.get(k)??new Set<T>(),seen).keys().next().value;
+        } while (k)
+    }
+
+    //"│├└┌┼╂┃┠┏┗┞┟↳↱→←"
+    //"│├└┌┼"
+    const cols:string[][] = [];
+    for (let l of listing){
+        const n = graph.get(l);
+        const li = indices.get(l);
+        if (n && (li !== undefined) && n.size > Number(Boolean(listing[li+1] && n.has(listing[li+1])))){
+            const col:string[] = [];
+            let minI = li;
+            let maxI = li;
+            for (let other of n){
+                const i = indices.get(other);
+                if (i !== undefined){
+                    minI = Math.min(i,minI);
+                    maxI = Math.max(i,maxI);
+                }
             }
-            // Check if all nodes this node must come after are already in the list.
-            if (subset(incomingEdges, usedNodes)) {
-                sorted.push(node);
-                usedNodes.add(node);
+            for (let i = 0; i < listing.length; i++){
+                const a = "  ┌┌│├└└  "[(Number(i >= minI) + Number(i > minI) + Number(i >= maxI) + Number(i > maxI))*2+Number(n.has(listing[i]) && i != li+1)]
+                const b = " ><x"[Number(n.has(listing[i]) && i != li+1)+2*Number(i==li)]
+                col.push(a+b);
             }
+            cols.push(col);
         }
-        if (sorted.length === lastLength) {
-            throw new Error('Graph contains a cycle');
+    }
+    let res = "";
+    for (let i = 0; i < listing.length; i++){
+        let line = "";
+        const l = listing[i];
+        const n = graph.get(l);
+        if (!(n && n.has(listing[i+1]))){
+            line = "\x1b[4m";
+        }
+        for (let j = 0; j < cols.length; j++){
+            line += cols[j][i];
+        }
+        res += "\n"+line + l + "\x1b[m";
+    }
+    return res;
+}
+
+function trimToCyclic<T>(graph: Map<T, Set<T>>): Map<T, Set<T>>{
+    const trimmed = new Map<T,Set<T>>();
+    const reverse = new Map<T,Set<T>>();
+    const domain = new Set<T>(graph.keys());
+    
+    for (let [k,v] of graph.entries()){
+        trimmed.set(k,setIntersection(v,domain));
+        for (let n of v){
+            if (domain.has(n)){
+                if (!reverse.has(n)){
+                    reverse.set(n,new Set<T>);
+                }
+                reverse.get(n)?.add(k);
+            }
+        }
+    }
+
+    const ends = setDifference(domain,new Set<T>(reverse.keys()));
+    for (let [k,v] of trimmed.entries()){
+        if (v.size == 0){
+            ends.add(k);
         }
     }
 
-    return sorted;
+    while (ends.size > 0){
```

Comment:
Please explain how this works. 
- Find nodes that have no incoming edges or no outgoing edges
- delete them
- repeat